src/http/multipart.rs

```diff
@@ -0,0 +1,98 @@
+use bytes::Buf;
+use reqwest::{
+    multipart::{Form, Part},
+    Client,
+    Url,
+};
+use tokio::{fs::File, io::AsyncReadExt};
+
+use super::AttachmentType;
+use crate::internal::prelude::*;
+
+/// Holder for multipart body. Contains files and payload_json for creating
+/// messages with attachments.
+#[derive(Clone, Debug)]
+pub struct Multipart<'a> {
+    pub files: Vec<AttachmentType<'a>>,
+    pub payload_json: JsonMap,
+}
+
+impl<'a> Multipart<'a> {
+    pub(crate) async fn to_multipart_form(&mut self, client: &Client) -> Result<Form> {
+        let mut multipart = Form::new();
+
+        for (file_num, file) in self.files.iter_mut().enumerate() {
+            match file {
+                AttachmentType::Bytes {
+                    data,
+                    filename,
+                } => {
+                    multipart = multipart.part(
+                        file_num.to_string(),
+                        Part::bytes(data.clone().into_owned()).file_name(filename.clone()),
+                    );
+                },
+                AttachmentType::File {
+                    file,
+                    filename,
+                } => {
+                    let mut buf = Vec::new();
+                    file.try_clone().await?.read_to_end(&mut buf).await?;
+
+                    multipart = multipart
+                        .part(file_num.to_string(), Part::stream(buf).file_name(filename.clone()));
+                },
+                AttachmentType::Path(path) => {
+                    let filename =
+                        path.file_name().map(|filename| filename.to_string_lossy().into_owned());
+                    let mut f = File::open(path).await?;
+                    let mut buf = vec![];
+                    f.read_to_end(&mut buf).await?;
+
+                    // Modify current AttachmentType as saved Bytes as to prevent
+                    // the need for another disk read when retrying
+                    *file = AttachmentType::Bytes {
+                        data: buf.clone().into(),
+                        filename: filename.clone().unwrap_or_else(String::new),
+                    };
```

Comment:
Not entirely sure if it is worth making an extra copy of attachment bytes every time an attachment is sent in order to not read from disk / make a network request in the case the message send fails or is rate limited.